File: Java/LeetCode/src/PlayWithAlgorithms/LazyPrimMST.java
package PlayWithAlgorithms;

import java.util.Vector;

/**
 * Created by Lix on 2017/9/18.
 */
public class LazyPrimMST<Weight extends Number & Comparable> {
    private WeightedGraph<Weight> G;
    private MinHeap<Edge<Weight>> pq;
    private boolean[] marked;
    private Vector<Edge<Weight>> mst;
    private Number mstWeight;

    public LazyPrimMST(WeightedGraph<Weight> graph) {
        G = graph;
        pq = new MinHeap<Edge<Weight>>(G.E());
        marked = new boolean[G.V()];
        mst = new Vector<Edge<Weight>>();

        visit(0);
        while (!pq.isEmpty()) {
            Edge<Weight> e = pq.extractMin();
            if (marked[e.v()] == marked[e.w()]) {
                continue;
            }

            mst.add(e);
            if (!marked[e.v()]) {
                visit(e.v());
            } else {
                visit(e.w());
            }
        }

        mstWeight = mst.elementAt(0).wt();
        for (int i = 1; i < mst.size(); i++) {
            mstWeight = mstWeight.doubleValue() + mst.elementAt(i).wt().doubleValue();
        }
    }

    private void visit(int v) {
        assert(!marked[v]);
        marked[v] = true;

        for (Edge<Weight> e : G.adj(v)) {
            if (!marked[e.other(v)]) {
                pq.insert(e);
            }
        }
    }

    Vector<Edge<Weight>> mstEdges() {
        return mst;
    }

    Number result() {
        return mstWeight;
    }
}

File: src/algs4/graph/lazy-prim-mst.ts
import { cloneDeep } from 'lodash'
import Queue from '../1-3/node-queue'
import MinPQ from '../sort/min-pq'
import Edge from './edge'
import EdgeWeightedGraph from './edge-weighted-graph'

/**
 *
 * 最小生成树 Prim 算法的延时实现
 *
 */
export default class LazyPrimMST {
  private marked: boolean[] // 最小生成树的顶点
  private mst: Queue<Edge> // 最小生成树的边
  private pq: MinPQ<Edge> // 横切边(包括失效的边)

  constructor(G: EdgeWeightedGraph) {
    this.pq = new MinPQ()
    this.marked = []
    this.mst = new Queue()

    this.visit(G, 0) // 假设 G 是连通的
    while (!this.pq.isEmpty()) {
      const e = this.pq.delMin() // 从 pq 中得到权重最小的边

      const v = e.either()
      const w = e.other(v)
      if (this.marked[v] && this.marked[w]) continue // 跳过失效的边
      this.mst.enqueue(e) // 将边添加到树中
      if (!this.marked[v]) this.visit(G, v) // 将顶点 v 或 w 添加到树中
      if (!this.marked[w]) this.visit(G, w)
    }
  }

  private visit(G: EdgeWeightedGraph, v: number) {
    // 标记顶点 v 并将所有连接 v 和未被标记顶点的边加入 pq
    this.marked[v] = true
    const adj = G.getAdj(v)
    while (adj.hasNext()) {
      const e = adj.next()
      if (!this.marked[e.other(v)]) this.pq.insert(e)
    }
  }

  getEdges() {
    return this.mst
  }

  getWeight() {
    let weight = 0.0
    const edges = cloneDeep(this.getEdges())
    while (!edges.isEmpty()) {
      weight += edges.dequeue().getWeight()
    }
    return weight
  }
}
